rfc/20250711-Middleware-For-Enhanced-Operations.md

```diff
@@ -0,0 +1,378 @@
+- **No feedback loop**: External tools cannot influence OpenTofu's behavior or provide metadata back to the state
+
+This RFC proposes bringing this extensibility directly into OpenTofu through a middleware system that allows users to intercept and augment operations at key points during execution. By providing official hook points and a well-defined protocol, we can enable the same policy enforcement, cost estimation, and compliance validation use cases while offering several advantages:
+
+- **Real-time validation**: Catch issues during planning, not after
+- **Rich context**: Access to full resource state, provider details, and operation context
+- **Simplified workflows**: Middleware runs as part of normal OpenTofu operations
+- **Bidirectional communication**: Middleware can fail operations and store metadata in state
+- **Language agnostic**: Write middleware in any language
+
+By offering this extensibility through a well-defined protocol, we aim to enable organizations to bring their existing policy and governance tools into the OpenTofu execution flow while fostering innovation in new use cases that weren't possible with external analysis alone.
+
+The architecture of this middleware system is inspired by the recent Model Context Protocol (MCP) approach, which has proven successful for extending AI assistants through local server processes communicating via JSON-RPC over stdio (along with other transports). This design pattern offers simplicity, language independence, and process isolation - qualities that align perfectly with OpenTofu's needs for a stable, extensible middleware system without the overhead of producing a full SDK or requiring middleware authors to be locked into a specific languge/ecosystem. You can find more information about this specification [at modelcontextprotocol.io/docs](https://modelcontextprotocol.io/docs/concepts/architecture).
+
+## Approach
+To provide users with a flexible and powerful way to extend OpenTofu's functionality, I propose implementing a middleware system that communicates with external processes via JSON-RPC 2.0 over STDIO. The middleware will be invoked at specific hook points during OpenTofu operations, allowing users to inspect, validate, and augment the behavior of resource operations without requiring changes to OpenTofu's core.
+
+### Hook Points for Operations
+The middleware system will provide hook points at two levels:
+
+**Resource-level hooks** (called for each individual resource/data source):
+- **pre-plan**
+- **post-plan**
+- **pre-apply**
+- **post-apply**
+- **pre-refresh**
+- **post-refresh**
+
+**Operation-level hooks** (called once per operation):
+- **init-stage-start**
+- **init-stage-complete**
+- **plan-stage-start**
+- **plan-stage-complete**
+- **apply-stage-start**
+- **apply-stage-complete**
+
+Additional hooks can be added in future versions as use cases emerge. It is possible that we should also consider hooks that handle failures/errors too.
```

Comment:
I know that recently there was a conversation about this: could we add also a "heartbeat" operation? Could be configurable with a  1m00s default.
Some could find that quite useful.

Reply:
What would the use-case be for writing custom middleware to happen on a "heartbeat" ?

Reply:
Mostly related to process monitors. Those systems can inject a middleware configuration at the project level and have custom logic around the heartbeat.
I am mostly referring to this: #2948